onadata/libs/utils/csv_builder.py

```diff
@@ -622,38 +610,44 @@ def export_to(self, path, dataview=None):
                                          filter_query=self.filter_query)
             if isinstance(cursor, QuerySet):
                 cursor = cursor.iterator()
-            self._update_columns_from_data(cursor)
+
+            self._update_ordered_columns_from_data(cursor)
+
+            data = self._format_for_dataframe(cursor)
 
             columns = list(chain.from_iterable(
                 [[xpath] if cols is None else cols
                  for (xpath, cols) in iteritems(self.ordered_columns)
                  if [c for c in dataview.columns if xpath.startswith(c)]]
             ))
-            cursor = dataview.query_data(dataview, all_data=True,
-                                         filter_query=self.filter_query)
-            if isinstance(cursor, QuerySet):
-                cursor = cursor.iterator()
-            data = self._format_for_dataframe(cursor)
         else:
-            cursor = self._query_data(self.filter_query)
+            try:
+                cursor = self._query_data(self.filter_query)
+            except NoRecordsFoundError:
+                # Set cursor object to an an empty queryset
+                cursor = self.xform.instances.none()
+                # Define export columns using xform schema
+                columns = self.xform.get_headers()
```

Comment:
Seems this variable isn't used anywhere at the moment and is reset at Line 642 ?

Reply:
This variable is passed to the `write_to_csv` function, but we initially define it here, containing a list of strings(column headers) to be used for the export.
While the variable is defined here, it is updated, not reset, on line 640 and again on line 645 with extra columns for the export

Reply:
I'm still unsure how line 640 updates the variable....

```
columns = list(chain.from_iterable(
                [[xpath] if cols is None else cols
                    for (xpath, cols) in iteritems(self.ordered_columns)]))
```

The list comprehension used there just seems to override the initial value and doesn't take into account the what the value of `columns` was at the beginning

Reply:
Ah! I've seen where the confusion here may be coming from. I have made some changes here in regards to this. Thanks!